File: src/mysqlcookbook/recipes/api/04_stmt/stmt.php
<?php
# stmt.php: demonstrate statement processing in PHP
# (without placeholders)

# Note: This does't bother printing getCode() for errors because
# getMessage() includes the error code.

require_once "Cookbook.php";

try
{
  $dbh = Cookbook::connect ();
}
catch (PDOException $e)
{
  die ("Cannot connect to server: " . $e->getMessage () . "\n");
}

# Default is same as PDO::FETCH_BOTH. Result can be used as ordered or
# associative array.

print ("Fetch using ordered array\n");
try
{
#@ _FETCHLOOP_1_
  $sth = $dbh->query ("SELECT id, name, cats FROM profile");
  $count = 0;
  while ($row = $sth->fetch (PDO::FETCH_NUM))
  {
    printf ("id: %s, name: %s, cats: %s\n", $row[0], $row[1], $row[2]);
    $count++;
  }
  printf ("Number of rows returned: %d\n", $count);
#@ _FETCHLOOP_1_
}
catch (PDOException $e)
{
  print ("Oops, the statement failed\n");
  print ("Error message: " . $e->getMessage () . "\n");
}

print ("Fetch using associative array\n");
try
{
  $sth = $dbh->query ("SELECT id, name, cats FROM profile");
#@ _FETCHLOOP_2_
  while ($row = $sth->fetch (PDO::FETCH_ASSOC))
    printf ("id: %s, name: %s, cats: %s\n",
            $row["id"], $row["name"], $row["cats"]);
#@ _FETCHLOOP_2_
}
catch (PDOException $e)
{
  print ("Oops, the statement failed\n");
  print ("Error message: " . $e->getMessage () . "\n");
}

print ("Fetch using object\n");
try
{
  $sth = $dbh->query ("SELECT id, name, cats FROM profile");
#@ _FETCHLOOP_3_
  while ($row = $sth->fetch (PDO::FETCH_OBJ))
    printf ("id: %s, name: %s, cats: %s\n", $row->id, $row->name, $row->cats);
#@ _FETCHLOOP_3_
}
catch (PDOException $e)
{
  print ("Oops, the statement failed\n");
  print ("Error message: " . $e->getMessage () . "\n");
}

print ("Fetch after setting to database handle to associative fetch mode\n");
try
{
#@ _DBH_SET_FETCH_MODE_
  $dbh->setAttribute (PDO::ATTR_DEFAULT_FETCH_MODE, PDO::FETCH_ASSOC);
#@ _DBH_SET_FETCH_MODE_
  $sth = $dbh->query ("SELECT id, name, cats FROM profile");
  while ($row = $sth->fetch ())
    # Use print_r to make it more clear what the row structure is
    print_r ($row);
}
catch (PDOException $e)
{
  print ("Oops, the statement failed\n");
  print ("Error message: " . $e->getMessage () . "\n");
}

print ("Fetch after setting to statement handle to object fetch mode\n");
try
{
  $sth = $dbh->query ("SELECT id, name, cats FROM profile");
#@ _STH_SET_FETCH_MODE_
  $sth->setFetchMode (PDO::FETCH_OBJ);
#@ _STH_SET_FETCH_MODE_
  while ($row = $sth->fetch ())
    # Use print_r to make it more clear what the row structure is
    print_r ($row);
}
catch (PDOException $e)
{
  print ("Oops, the statement failed\n");
  print ("Error message: " . $e->getMessage () . "\n");
}

print ("Fetch rows using statement handle as iterator\n");
try
{
  $sth = $dbh->query ("SELECT id, name, cats FROM profile");
#@ _FETCH_ITERATOR_1_
  $sth->setFetchMode (PDO::FETCH_NUM);
  foreach ($sth as $row)
    printf ("id: %s, name: %s, cats: %s\n", $row[0], $row[1], $row[2]);
#@ _FETCH_ITERATOR_1_
}
catch (PDOException $e)
{
  print ("Oops, the statement failed\n");
  print ("Error message: " . $e->getMessage () . "\n");
}

# Default is same as PDO::FETCH_BOTH. Result can be used as ordered or
# associative array.

print ("Fetch using fetchAll()\n");
try
{
  $sth = $dbh->query ("SELECT id, name, cats FROM profile");
#@ _FETCHALL_LOOP_
  $rows = $sth->fetchAll (PDO::FETCH_NUM);
  foreach ($rows as $row)
    printf ("id: %s, name: %s, cats: %s\n", $row[0], $row[1], $row[2]);
#@ _FETCHALL_LOOP_
}
catch (PDOException $e)
{
  print ("Oops, the statement failed\n");
  print ("Error message: " . $e->getMessage () . "\n");
}

print ("Update rows\n");
try
{
#@ _UPDATE_
  $count = $dbh->exec ("UPDATE profile SET cats = cats+1 WHERE name = 'Sybil'");
  printf ("Number of rows updated: %d\n", $count);
#@ _UPDATE_
}
catch (PDOException $e)
{
  print ("Oops, the statement failed\n");
  print ("Error message: " . $e->getMessage () . "\n");
}

$dbh = NULL;
?>

File: src/mysqlcookbook/recipes/api/04_stmt/stmt.pl
#!/usr/bin/perl
# stmt.pl: demonstrate statement processing in Perl
# (without placeholders)

use strict;
use warnings;
use Cookbook;

my $dbh = Cookbook::connect ();

{ # begin scope

# NOTE: THIS EXAMPLE MUST INCLUDE finish() (it's discussed in the text)

print "Fetch rows with fetchrow_array()\n";
#@ _FETCHLOOP_1_
my $sth = $dbh->prepare ("SELECT id, name, cats FROM profile");
$sth->execute ();
my $count = 0;
while (my @val = $sth->fetchrow_array ())
{
  print "id: $val[0], name: $val[1], cats: $val[2]\n";
  ++$count;
}
$sth->finish ();
print "Number of rows returned: $count\n";
#@ _FETCHLOOP_1_

} # end scope

{ # begin scope

print "Fetch rows with fetchrow_arrayref()\n";
my $sth = $dbh->prepare ("SELECT id, name, cats FROM profile");
$sth->execute ();
#@ _FETCHLOOP_2_
while (my $ref = $sth->fetchrow_arrayref ())
{
  print "id: $ref->[0], name: $ref->[1], cats: $ref->[2]\n";
}
#@ _FETCHLOOP_2_

} # end scope

{ # begin scope

print "Fetch rows with fetchrow_hashref()\n";
my $sth = $dbh->prepare ("SELECT id, name, cats FROM profile");
$sth->execute ();
#@ _FETCHLOOP_3_
while (my $ref = $sth->fetchrow_hashref ())
{
  print "id: $ref->{id}, name: $ref->{name}, cats: $ref->{cats}\n";
}
#@ _FETCHLOOP_3_

} # end scope

{ # begin scope

print "Fetch rows with selectrow_array()\n";
#@ _SELECTROW_ARRAY_1_
my @val = $dbh->selectrow_array ("SELECT name, birth, foods FROM profile
                                  WHERE id = 3");
#@ _SELECTROW_ARRAY_1_
#@ _SELECTROW_ARRAY_2_
my $buddy_count = $dbh->selectrow_array ("SELECT COUNT(*) FROM profile");
#@ _SELECTROW_ARRAY_2_
print "buddy row: @val\n";
print "buddy count: $buddy_count\n";

} # end scope

{ # begin scope

print "Update rows\n";
#@ _DO_1_
my $count = $dbh->do ("UPDATE profile SET cats = cats+1
                       WHERE name = 'Sybil'");
if ($count)   # print row count if no error occurred
{
  $count += 0;
  print "Number of rows updated: $count\n";
}
#@ _DO_1_

} # end scope

{ # begin scope

my $count = $dbh->do ("UPDATE profile SET cats = cats+1
                       WHERE name = 'Sybil'");
#@ _DO_2_
if ($count)   # print row count if no error occurred
{
  printf "Number of rows updated: %d\n", $count;
}
#@ _DO_2_

} # end scope

{ # begin scope

#@ _DO_3_
my $count = $dbh->do ("UPDATE profile SET cats = cats+1
                       WHERE name = 'Sybil'");
printf "Number of rows updated: %d\n", $count;
#@ _DO_3_

} # end scope

$dbh->disconnect ();
